Refine tail call conditions for `tail` and `musttail` markers

antoniofrighetto opened the pull request:
musttail has stronger requirements than tail, so ensure this is taken into account.

nunoplopes commented on tests/alive-tv/calls/tailcall-memcpy.srctgt.ll:
why delete this test?  it's correct

antoniofrighetto replied:
Moved it to `musttail-memcpy.srctgt.ll` and replace tail w/ musttail (since with tail only marker it shouldn't be UB, we're verifying only isMustTailCall now).

nunoplopes commented on tests/alive-tv/calls/musttail-memcpy.srctgt.ll:
```diff
@@ -0,0 +1,18 @@
+  %call2 = alloca i64
+  %call3 = alloca i64
+  store i64 %arg, ptr %call2, align 8
+  musttail call tailcc void @llvm.memcpy.p0.p0.i64(ptr byval(ptr) %call3, ptr byval(ptr) %call2, i64 8, i1 false)
```
this not a good test because it's mixing several conditions at once

antoniofrighetto replied:
Separated it in multiple functions.

antoniofrighetto commented:
I think we miss checking the following:
> In addition, if the calling convention is not `swifttailcc` or `tailcc`:
      - All ABI-impacting function attributes, such as sret, byval, inreg,
   returned, and inalloca, must match.
      - The caller and callee prototypes must match. Pointer types of parameters
   or return types may differ in pointee type, but not in address space.

(Not completely sure if this refers to musttail too though)

nunoplopes commented:
From a quick read, this doesn't seem to be what we want at all. We need both tail & musttail, as both have constraints.
I suggest create a new thing in ir/attr.h (tail call type or smth like that). It can then encode the conditions for each case.

antoniofrighetto commented:
Added a new type `TailCallInfo` to encode both tail call markers. Updated the preconditions such that they are verified for both markers: tail position only for musttail, whereas callee not accessing allocas, va_args, or byval arguments from the caller for both. Restored `tailcall-memcpy.srctgt.ll` test since it violates this last constraint.

nunoplopes commented on ir/instr.cpp:
```diff
@@ -2,6 +2,7 @@
 // Distributed under the MIT license that can be found in the LICENSE file.
 
 #include "ir/instr.h"
+#include "ir/attrs.h"
```
not needed

nunoplopes commented on llvm_util/utils.cpp:
```diff
@@ -521,4 +525,67 @@ llvm::Function *findFunction(llvm::Module &M, const string &FName) {
   return F && !F->isDeclaration() ? F : nullptr;
 }
 
+TailCallInfo parse_tail_call_info(const llvm::CallInst &i) {
```
we don't encode semantics like this. The semantics must be in the ir folder.
We support other IRs -> Alive2 IR, hence we do as little as possible in the translation.

antoniofrighetto replied:
I left `parse_tail_call_info` in utils.cpp so that it may be used from known_functions too (originally was in llvm2alive). I think we still need it there as we're parsing data from LLVM (like `parse_fn_attrs`). Do you mean that the semantics in `are_tailcall_preconditions_met` is to be moved in ir inside `check_tailcall`?

antoniofrighetto replied:
Also made it more coherent (`parse_tail_call_info` -> `parse_fn_tailcall`).

nunoplopes commented on llvm_util/known_fns.h:
```diff
@@ -18,6 +18,7 @@ class FnAttrs;
 class Instr;
 class ParamAttrs;
 class Value;
+struct TailCallInfo;
```
seeems unused

nunoplopes commented on llvm_util/known_fns.cpp:
```diff
@@ -520,27 +520,32 @@ known_call(llvm::CallInst &i, const llvm::TargetLibraryInfo &TLI,
   case llvm::LibFunc_memset: // void* memset(void *ptr, int val, size_t bytes)
-    BB.addInstr(make_unique<Memset>(*args[0], *args[1], *args[2], 1, is_tailcall));
+    BB.addInstr(make_unique<Memset>(*args[0], *args[1], *args[2], 1,
+                                    std::move(tci)));
```
move is a bit overkill

antoniofrighetto replied:
To be changed after the change above.

nunoplopes commented on ir/instr.cpp:
```diff
@@ -2314,28 +2314,36 @@ void FnCall::print(ostream &os) const {
   os << ')' << attrs;
 }
 
-static void check_tailcall(const Instr &i, State &s) {
```
you can move this function into the tailcallinfo struct

antoniofrighetto commented on tests/alive-tv/calls/musttail-call.srctgt.ll:
```diff
@@ -0,0 +1,11 @@
+define tailcc i64 @src(i64 noundef %arg) {
```
Dropped it.

antoniofrighetto commented:
Updated, moved semantics in checkTailCall.
